File: src/main/java/com/ampro/weebot/commands/properties/Restriction.java
package com.ampro.weebot.commands.properties;

import com.ampro.weebot.Launcher;
import net.dv8tion.jda.core.EmbedBuilder;
import net.dv8tion.jda.core.entities.*;
import net.dv8tion.jda.core.managers.GuildController;

import javax.annotation.Nonnull;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

/**
 * Contains validation information and methods to restrict access to any entity.
 * <br>
 *     Restrictions by positivity (disallow or allow access). <br>
 *     Restrictions can be made by {@link Role}, {@link TextChannel}, or
 *     {@link User}. <br>
 *     Restrictions are made using
 *     {@link net.dv8tion.jda.core.entities.ISnowflake} Long IDs. <br>
 *
 *
 * @author Jonathan Augustine
 *
 */
public class Restriction {

    /** An indication of the Command's restriction in a guild. */
    public enum Status {
        /** The Command has no restrictions */
        OPEN,
        /** The Command has restrictions */
        RESTRICTED,
        /** The Command is disabled */
        DISABLED
    }

    private final List<Long> allowedUsers;
    private final List<Long> blockedUsers;
    private final List<Long> allowedRoles;
    private final List<Long> blockedRoles;
    private final List<Long> allowedTextChannels;
    private final List<Long> blockedTextChannels;

    /** Initialize all restriction arrays */
    public Restriction() {
        this.allowedUsers = new ArrayList<>();
        this.blockedUsers = new ArrayList<>();
        this.allowedRoles = new ArrayList<>();
        this.blockedRoles = new ArrayList<>();
        this.allowedTextChannels = new ArrayList<>();
        this.blockedTextChannels = new ArrayList<>();
    }

    /**
     * Add allowed {@link User users}, {@link Role roles}, or
     * {@link TextChannel TextChannels}.
     * Will remove any from the "blocked" list if present.
     * @param iSnowflakes The iSnowflakes to allow
     */
    public void allow(@Nonnull List<? extends ISnowflake> iSnowflakes) {
        if (iSnowflakes.isEmpty()) {
            return;
        }
        if (iSnowflakes.get(0) instanceof User) {
            iSnowflakes.forEach( user -> {
                this.blockedUsers.remove(user.getIdLong());
                this.allowedUsers.add(user.getIdLong());
            });
        } else if (iSnowflakes.get(0) instanceof Role) {
            iSnowflakes.forEach( role -> {
                this.blockedRoles.remove(role.getIdLong());
                this.allowedRoles.add(role.getIdLong());
            });
        } else if (iSnowflakes.get(0) instanceof TextChannel) {
            iSnowflakes.forEach( channel -> {
                this.blockedTextChannels.remove(channel.getIdLong());
                this.allowedTextChannels.add(channel.getIdLong());
            });
        }
    }

    /**
     * Add blocked {@link User users}, {@link Role roles}, or
     * {@link TextChannel TextChannels}.
     * Will remove any from the "allowed" list if present.
     * @param iSnowflakes The iSnowflakes to block
     */
    public void block(@Nonnull List<? extends ISnowflake> iSnowflakes) {
        if (iSnowflakes.isEmpty()) {
            return;
        }
        if (iSnowflakes.get(0) instanceof User) {
            iSnowflakes.forEach( user -> {
                this.allowedUsers.remove(user.getIdLong());
                this.blockedUsers.add(user.getIdLong());
            });
        } else if (iSnowflakes.get(0) instanceof Role) {
            iSnowflakes.forEach( role -> {
                this.allowedRoles.remove(role.getIdLong());
                this.blockedRoles.add(role.getIdLong());
            });
        } else if (iSnowflakes.get(0) instanceof TextChannel) {
            iSnowflakes.forEach( channel -> {
                this.allowedTextChannels.remove(channel.getIdLong());
                this.blockedTextChannels.add(channel.getIdLong());
            });
        }
    }

    /** @return An {@link Collections#unmodifiableList(List)} of allowed users. */
    public List<Long> getAllowedUsers() {
        return Collections.unmodifiableList(allowedUsers);
    }

    /** @return An {@link Collections#unmodifiableList(List)} of blocked users. */
    public List<Long> getBlockedUsers() {
        return Collections.unmodifiableList(blockedUsers);
    }

    /**
     * Add an allowed {@link User}.
     * Will remove the user from the "blocked" list if present.
     * @param user The user to allow
     */
    public void allow(User user) {
        this.blockedUsers.remove(user.getIdLong());
        this.allowedUsers.add(user.getIdLong());
    }

    /**
     * Block a user's access.
     * Will remove the user from "allowed" list if present
     * @param user The user to black access from
     */
    public void block(User user) {
        this.allowedUsers.remove(user.getIdLong());
        this.blockedUsers.add(user.getIdLong());
    }

    /**
     * Check if a user has been allowed or blocked.
     * @param user The user to check
     * @return {@code true} if the user is allowed, or there are no restrictions.
     */
    public boolean isAllowed(User user) {
        if (!this.allowedUsers.isEmpty()) {
            return this.allowedUsers.contains(user.getIdLong());
        } else if (!this.blockedUsers.isEmpty()) {
            return !this.blockedUsers.contains(user.getIdLong());
        } else {
            return true;
        }
    }

    /** @return An {@link Collections#unmodifiableList(List)} of allowed Roles. */
    public List<Long> getAllowedRoles() {
        return Collections.unmodifiableList(allowedRoles);
    }

    /** @return An {@link Collections#unmodifiableList(List)} of blocked Roles. */
    public List<Long> getBlockedRoles() {
        return Collections.unmodifiableList(blockedRoles);
    }

    /**
     * Allow a Role access.
     * Will remove the role from the "blocked" list if present.
     * @param role The role to allow access
     */
    public void allow(Role role) {
        this.blockedRoles.remove(role.getIdLong());
        this.allowedRoles.add(role.getIdLong());
    }

    /**
     * Block a Role's access.
     * Will remove the Role from the "blocked" list if present.
     * @param role The role to block access.
     */
    public void block(Role role) {
        this.allowedRoles.remove(role.getIdLong());
        this.blockedRoles.add(role.getIdLong());
    }

    /**
     * Check if a Role has been allowed or blocked.
     * @param role The user to check
     * @return {@code true} if the role is allowed, or there are no restrictions.
     */
    public boolean isAllowed(Role role) {
        if (!this.allowedRoles.isEmpty()) {
            return this.allowedRoles.contains(role.getIdLong());
        } else if (!this.blockedRoles.isEmpty()) {
            return !this.blockedRoles.contains(role.getIdLong());
        } else {
            return true;
        }
    }

    /** @return An {@link Collections#unmodifiableList(List)} of allowed TextChannels. */
    public List<Long> getAllowedTextChannels() {
        return Collections.unmodifiableList(allowedTextChannels);
    }

    /** @return An {@link Collections#unmodifiableList(List)} of blocked TextChannels. */
    public List<Long> getBlockedTextChannels() {
        return Collections.unmodifiableList(blockedTextChannels);
    }

    /**
     * Allow a TextChannel access.
     * Will remove the channel from the "blocked" list if present.
     * @param channel The TextChannel to allow access
     */
    public void allow(TextChannel channel) {
        this.blockedTextChannels.remove(channel.getIdLong());
        this.allowedTextChannels.add(channel.getIdLong());
    }

    /**
     * Block a TextChannel access.
     * Will remove the channel from the "allowed" list if present.
     * @param channel The TextChannel to block access
     */
    public void block(TextChannel channel) {
        this.allowedTextChannels.remove(channel.getIdLong());
        this.blockedTextChannels.add(channel.getIdLong());
    }

    /**
     * Check if a TextChannel has been allowed or blocked.
     * @param channel The TextChannel to check
     * @return {@code true} if the TextChannel is allowed,
     *          or there are no restrictions.
     */
    public boolean isAllowed(TextChannel channel) {
        if (!this.allowedTextChannels.isEmpty()) {
            return this.allowedTextChannels.contains(channel.getIdLong());
        } else if (!this.blockedTextChannels.isEmpty()) {
            return !this.blockedTextChannels.contains(channel.getIdLong());
        } else {
            return true;
        }
    }

    /** @return {@code true} if there are any restrictions.*/
    public boolean restricted() {
        return allowedUsers.isEmpty() || blockedUsers.isEmpty()
                || allowedRoles.isEmpty() || blockedRoles.isEmpty()
                || allowedTextChannels.isEmpty() || blockedTextChannels .isEmpty();
    }

    /**
     * Get an {@link EmbedBuilder} with each allow or block list as it's own
     * field. The EmbedBuilder is in standard Weebot form, untitled with no
     * description.
     * @param guild The guild the restriction is housed in.
     * @return
     */
    public EmbedBuilder toEmbedBuilder(Guild guild) {
        StringBuilder sb = new StringBuilder();
        EmbedBuilder eb = Launcher.getStandardEmbedBuilder();
        if (!this.getAllowedUsers().isEmpty()) {
            this.getAllowedUsers().forEach(u -> {
                sb.append("*")
                  .append(guild.getMemberById(u).getEffectiveName())
                  .append("*\n");
            });
            eb.addField("Allowed Members", sb.toString(), true);
            sb.setLength(0);
        }
        if (!this.getAllowedRoles().isEmpty()) {
            this.getAllowedRoles().forEach(r -> {
                sb.append("*")
                  .append(guild.getRoleById(r).getName())
                  .append("*\n");
            });
            eb.addField("Allowed Roles", sb.toString(), true);
            sb.setLength(0);
        }
        if (!this.getAllowedTextChannels().isEmpty()) {
            this.getAllowedTextChannels().forEach(tc -> {
                sb.append("*")
                  .append(guild.getTextChannelById(tc).getName())
                  .append("*\n");
            });
            eb.addField("Allowed TextChannels", sb.toString(), true);
            sb.setLength(0);
        }
        if (!this.getBlockedUsers().isEmpty()) {
            this.getBlockedUsers().forEach(u -> {
                sb.append("*")
                  .append(guild.getMemberById(u).getEffectiveName())
                  .append("*\n");
            });
            eb.addField("Blocked Members", sb.toString(), true);
            sb.setLength(0);
        }
        if (!this.getBlockedRoles().isEmpty()) {
            this.getBlockedRoles().forEach(r -> {
                sb.append("*")
                  .append(guild.getRoleById(r).getName())
                  .append("*\n");
            });
            eb.addField("BlockedRoles", sb.toString(), true);
            sb.setLength(0);
        }
        if (!this.getBlockedTextChannels().isEmpty()) {
            this.getBlockedTextChannels().forEach(tc -> {
                sb.append("*")
                  .append(guild.getTextChannelById(tc).getName())
                  .append("*\n");
            });
            eb.addField("Blocked TextChannels", sb.toString(), true);
            sb.setLength(0);
        }
        return eb;
    }

}

File: src/main/kotlin/com/ampro/weebot/Restriction.kt
/*
 * Copyright Aquatic Mastery Productions (c) 2018.
 */

package com.ampro.weebot

import com.ampro.weebot.extensions.strdEmbedBuilder
import net.dv8tion.jda.core.EmbedBuilder
import net.dv8tion.jda.core.entities.*

/**
 * Contains validation information and methods to restrict access to any entity.
 * <br></br>
 * Restrictions by positivity (disallow or allow access). <br></br>
 * Restrictions can be made by [Role], [TextChannel], or
 * [User]. <br></br>
 * Restrictions are made using
 * [net.dv8tion.jda.core.entities.ISnowflake] Long IDs. <br></br>
 *
 *
 * @author Jonathan Augustine
 */
class Restriction {

    /** An indication of the Command's restriction in a guild.  */
    enum class Status {
        /** The Item has no restrictions  */
        OPEN,
        /** The Item has restrictions  */
        RESTRICTED,
        /** The Item is disabled  */
        DISABLED
    }

    val allowedUsers: MutableList<Long> = ArrayList()
    val blockedUsers: MutableList<Long> = ArrayList()
    val allowedRoles: MutableList<Long> = ArrayList()
    val blockedRoles: MutableList<Long> = ArrayList()
    val allowedTextChannels: MutableList<Long> = ArrayList()
    val blockedTextChannels: MutableList<Long> = ArrayList()
    val allowedVoiceChannels: MutableList<Long> = ArrayList()
    val blockedVoiceChannels: MutableList<Long> = ArrayList()

    /**
     * Add allowed [users][User], [roles][Role], or
     * [TextChannels][TextChannel].
     * Will remove any from the "blocked" list if present.
     * @param iSnowflakes The iSnowflakes to allow
     */
    fun allow(iSnowflakes: List<ISnowflake>) {
        if (iSnowflakes.isEmpty()) { return }
        when {
            iSnowflakes[0] is User -> iSnowflakes.forEach { user ->
                this.blockedUsers.remove(user.idLong)
                this.allowedUsers.add(user.idLong)
            }
            iSnowflakes[0] is Role -> iSnowflakes.forEach { role ->
                this.blockedRoles.remove(role.idLong)
                this.allowedRoles.add(role.idLong)
            }
            iSnowflakes[0] is TextChannel -> iSnowflakes.forEach { channel ->
                this.blockedTextChannels.remove(channel.idLong)
                this.allowedTextChannels.add(channel.idLong)
            }
            iSnowflakes[0] is VoiceChannel -> iSnowflakes.forEach { channel ->
                blockedVoiceChannels.remove(channel.idLong)
                allowedVoiceChannels.add(channel.idLong)
            }
        }
    }

    /**
     * Add blocked [users][User], [roles][Role], or
     * [TextChannels][TextChannel].
     * Will remove any from the "allowed" list if present.
     *
     * @param iSnowflakes The iSnowflakes to block
     */
    fun block(iSnowflakes: List<ISnowflake>) {
        if (iSnowflakes.isEmpty()) {
            return
        }
        when {
            iSnowflakes[0] is User -> iSnowflakes.forEach { user ->
                this.allowedUsers.remove(user.idLong)
                this.blockedUsers.add(user.idLong)
            }
            iSnowflakes[0] is Role -> iSnowflakes.forEach { role ->
                this.allowedRoles.remove(role.idLong)
                this.blockedRoles.add(role.idLong)
            }
            iSnowflakes[0] is TextChannel -> iSnowflakes.forEach { channel ->
                this.allowedTextChannels.remove(channel.idLong)
                this.blockedTextChannels.add(channel.idLong)
            }
        }
    }

    /**
     * Add an allowed [User].
     * Will remove the user from the "blocked" list if present.
     * @param user The user to allow
     */
    fun allow(user: User) {
        this.blockedUsers.remove(user.idLong)
        this.allowedUsers.add(user.idLong)
    }

    /**
     * Block a user's access.
     * Will remove the user from "allowed" list if present
     * @param user The user to black access from
     */
    fun block(user: User) {
        this.allowedUsers.remove(user.idLong)
        this.blockedUsers.add(user.idLong)
    }

    /**
     * Check if a user has been allowed or blocked.
     * @param user The user to check
     * @return `true` if the user is allowed, or there are no restrictions.
     */
    fun isAllowed(user: User): Boolean {
        return if (!this.allowedUsers.isEmpty()) {
            this.allowedUsers.contains(user.idLong)
        } else if (!this.blockedUsers.isEmpty()) {
            !this.blockedUsers.contains(user.idLong)
        } else {
            true
        }
    }

    /**
     * Allow a Role access.
     * Will remove the role from the "blocked" list if present.
     * @param role The role to allow access
     */
    fun allow(role: Role) {
        this.blockedRoles.remove(role.idLong)
        this.allowedRoles.add(role.idLong)
    }

    /**
     * Block a Role's access.
     * Will remove the Role from the "blocked" list if present.
     * @param role The role to block access.
     */
    fun block(role: Role) {
        this.allowedRoles.remove(role.idLong)
        this.blockedRoles.add(role.idLong)
    }

    /**
     * Check if a Role has been allowed or blocked.
     * @param role The user to check
     * @return `true` if the role is allowed, or there are no restrictions.
     */
    fun isAllowed(role: Role): Boolean {
        return if (!this.allowedRoles.isEmpty()) {
            this.allowedRoles.contains(role.idLong)
        } else if (!this.blockedRoles.isEmpty()) {
            !this.blockedRoles.contains(role.idLong)
        } else {
            true
        }
    }

    /**
     * Allow a TextChannel access.
     * Will remove the channel from the "blocked" list if present.
     * @param channel The TextChannel to allow access
     */
    fun allow(channel: TextChannel) {
        this.blockedTextChannels.remove(channel.idLong)
        this.allowedTextChannels.add(channel.idLong)
    }

    /**
     * Block a TextChannel access.
     * Will remove the channel from the "allowed" list if present.
     * @param channel The TextChannel to block access
     */
    fun block(channel: TextChannel) {
        this.allowedTextChannels.remove(channel.idLong)
        this.blockedTextChannels.add(channel.idLong)
    }

    /**
     * Check if a TextChannel has been allowed or blocked.
     * @param channel The TextChannel to check
     * @return `true` if the TextChannel is allowed,
     * or there are no restrictions.
     */
    fun isAllowed(channel: TextChannel): Boolean {
        return if (!this.allowedTextChannels.isEmpty()) {
            this.allowedTextChannels.contains(channel.idLong)
        } else if (!this.blockedTextChannels.isEmpty()) {
            !this.blockedTextChannels.contains(channel.idLong)
        } else {
            true
        }
    }

    /** @return `true` if there are any restrictions.
     */
    fun restricted(): Boolean {
        return (allowedUsers.isEmpty() || blockedUsers.isEmpty() || allowedRoles.isEmpty() || blockedRoles.isEmpty() || allowedTextChannels.isEmpty() || blockedTextChannels.isEmpty())
    }

    /**
     * Get an [EmbedBuilder] with each allow or block list as it's own
     * field. The EmbedBuilder is in standard Weebot form, untitled with no
     * description.
     * @param guild The guild the restriction is housed in.
     * @return
     */
    fun toEmbedBuilder(guild: Guild): EmbedBuilder {
        val sb = StringBuilder()
        val eb = strdEmbedBuilder
        if (!allowedUsers.isEmpty()) {
            this.allowedUsers.forEach { u ->
                sb.append("*").append(guild.getMemberById(u).effectiveName).append("*\n")
            }
            eb.addField("Allowed Members", sb.toString(), true)
            sb.setLength(0)
        }
        if (!this.allowedRoles.isEmpty()) {
            this.allowedRoles.forEach { r ->
                sb.append("*").append(guild.getRoleById(r).name).append("*\n")
            }
            eb.addField("Allowed Roles", sb.toString(), true)
            sb.setLength(0)
        }
        if (!this.allowedTextChannels.isEmpty()) {
            this.allowedTextChannels.forEach { tc ->
                sb.append("*").append(guild.getTextChannelById(tc).name).append("*\n")
            }
            eb.addField("Allowed TextChannels", sb.toString(), true)
            sb.setLength(0)
        }
        if (!this.blockedUsers.isEmpty()) {
            this.blockedUsers.forEach { u ->
                sb.append("*").append(guild.getMemberById(u).effectiveName).append("*\n")
            }
            eb.addField("Blocked Members", sb.toString(), true)
            sb.setLength(0)
        }
        if (!this.blockedRoles.isEmpty()) {
            this.blockedRoles.forEach { r ->
                sb.append("*").append(guild.getRoleById(r).name).append("*\n")
            }
            eb.addField("BlockedRoles", sb.toString(), true)
            sb.setLength(0)
        }
        if (!this.blockedTextChannels.isEmpty()) {
            this.blockedTextChannels.forEach { tc ->
                sb.append("*").append(guild.getTextChannelById(tc).name).append("*\n")
            }
            eb.addField("Blocked TextChannels", sb.toString(), true)
            sb.setLength(0)
        }
        return eb
    }

}
